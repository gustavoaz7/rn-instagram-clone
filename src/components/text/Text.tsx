import React from 'react';
import { Text as RNText, TextProps as RNTextProps } from 'react-native';
import styled from 'styled-components/native';

export interface TextProps extends RNTextProps {
  children?: React.ReactNode;
}

export function Text({ children, ...props }: TextProps): JSX.Element {
  return <StyledText {...props}>{children}</StyledText>;
}

const StyledText = styled(RNText)`
  color: ${({ theme }) => theme.color.black};
`;
