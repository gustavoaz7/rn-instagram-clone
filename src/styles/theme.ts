import type { Animated } from 'react-native';

const HEART_SPRING_SCALE_CONFIG: Animated.SpringAnimationConfig = {
  toValue: 1,
  useNativeDriver: true,
  velocity: 5,
  tension: 100,
  friction: 5,
};

export enum THEME_VARIANTS {
  LIGHT = 'light',
  DARK = 'dark',
}

export type TTheme = {
  spacing: {
    xs: string;
    s: string;
    m: string;
    l: string;
    xl: string;
  };
  color: {
    lightBlue: string;
    royalBlue: string;
    blue: string;
    purple: string;
    darkPink: string;
    purpleRed: string;
    red: string;
    darkOrange: string;
    orange: string;
    yellow: string;
    lightYellow: string;
    white: string;
    black: string;
    gray: string;
    foreground: string;
    background: string;
  };
  font: {
    size: {
      s: string;
      m: string;
      l: string;
      xl: string;
    };
  };
  animation: {
    timingFast: number;
    timingBase: number;
    timingSlow: number;
    heart: {
      initialScale: number;
      springConfig: Animated.SpringAnimationConfig;
    };
  };
};

const defaultSpacing: TTheme['spacing'] = {
  xs: '5px',
  s: '10px',
  m: '15px',
  l: '20px',
  xl: '25px',
};

const defaultColors: Omit<TTheme['color'], 'foreground' | 'background'> = {
  lightBlue: '#0095f6',
  royalBlue: '#405DE6',
  blue: '#5B51D8',
  purple: '#833AB4',
  darkPink: '#C13584',
  purpleRed: '#E1306C',
  red: '#FD1D1D',
  darkOrange: '#F56040',
  orange: '#F77737',
  yellow: '#FCAF45',
  lightYellow: '#FFDC80',
  white: '#FFFFFF',
  black: '#020202',
  gray: '#8e8e8e',
};

const defaultFont: TTheme['font'] = {
  size: {
    s: '12px',
    m: '14px',
    l: '16px',
    xl: '20px',
  },
};

const defaultAnimation: TTheme['animation'] = {
  timingFast: 150,
  timingBase: 300,
  timingSlow: 500,
  heart: {
    initialScale: 0.3,
    springConfig: HEART_SPRING_SCALE_CONFIG,
  },
};

export const theme: Record<THEME_VARIANTS, TTheme> = {
  [THEME_VARIANTS.LIGHT]: {
    spacing: defaultSpacing,
    color: {
      ...defaultColors,
      foreground: defaultColors.black,
      background: defaultColors.white,
    },
    font: defaultFont,
    animation: defaultAnimation,
  },
  [THEME_VARIANTS.DARK]: {
    spacing: defaultSpacing,
    color: {
      ...defaultColors,
      foreground: defaultColors.white,
      background: defaultColors.black,
    },
    font: defaultFont,
    animation: defaultAnimation,
  },
};
