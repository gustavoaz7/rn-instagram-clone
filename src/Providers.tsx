import React, { ReactNode } from 'react';
import { ThemeProvider } from 'styled-components/native';
import { Provider, ProviderProps } from 'react-redux';
import { PersistGate } from 'redux-persist/integration/react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { RootSiblingParent } from 'react-native-root-siblings'; // dep of rn-root-toast
import { theme as stylesTheme } from './styles/theme';
import { persistor, store as reduxStore } from './redux/store';
import { useThemeVariantSelector } from './redux/theme-variant';

type TProvidersProps = {
  children?: ReactNode;
  store?: ProviderProps['store'];
  theme?: typeof stylesTheme;
};

export function Providers({
  children = null,
  store = reduxStore,
  theme = stylesTheme,
}: TProvidersProps): JSX.Element {
  return (
    <RootSiblingParent>
      <Provider store={store}>
        <PersistGate loading={null} persistor={persistor}>
          <Theming theme={theme}>{children}</Theming>
        </PersistGate>
      </Provider>
    </RootSiblingParent>
  );
}

function Theming({
  theme,
  children,
}: Required<Omit<TProvidersProps, 'store'>>): JSX.Element {
  const themeVariant = useThemeVariantSelector();
  return <ThemeProvider theme={theme[themeVariant]}>{children}</ThemeProvider>;
}
