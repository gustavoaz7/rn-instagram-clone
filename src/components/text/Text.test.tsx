import React from 'react';
import { render, RenderOptions } from '@testing-library/react-native';
import { ThemeProvider } from 'styled-components/native';
import { Text, TextProps } from './Text';
import { Providers } from '../../Providers';
import { theme, THEME_VARIANTS } from '../../styles/theme';

describe('components - Text', () => {
  const options: RenderOptions = { wrapper: Providers };
  const text = 'Hello world';

  it('renders', () => {
    render(<Text>Hello</Text>, options);
  });

  it('matches snapshot', () => {
    const { toJSON } = render(<Text>Hello</Text>, options);

    expect(toJSON()).toMatchSnapshot();
  });

  it.each([THEME_VARIANTS.LIGHT, THEME_VARIANTS.DARK])(
    'uses foreground color for %p theme',
    variant => {
      const { getByText } = render(
        <ThemeProvider theme={theme[variant]}>
          <Text>{text}</Text>
        </ThemeProvider>,
        options,
      );

      expect(getByText(text)).toHaveStyle({
        color: theme[variant].color.foreground,
      });
    },
  );

  it('passes props to RN text component', () => {
    const props: TextProps = {
      lineBreakMode: 'clip',
      allowFontScaling: false,
      numberOfLines: 1,
    };
    const { getByText } = render(<Text {...props}>{text}</Text>, options);

    const element = getByText(text);

    expect(element).toBeTruthy();
    Object.entries(props).forEach(([key, value]) => {
      expect(element.props[key]).toEqual(value);
    });
  });
});
