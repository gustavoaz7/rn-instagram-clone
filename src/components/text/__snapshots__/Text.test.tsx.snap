// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components - Text matches snapshot 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#020202",
      },
    ]
  }
>
  Hello
</Text>
`;
